File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    int dfs(int r,int c,vector<vector<int>>& dp,vector<vector<int>>& matrix) {    
    if (dp[r][c]!=-1) return dp[r][c];
    vector<pair<int,int>> disp = {{1,0},{-1,0},{0,1},{0,-1}};
    int ans = -1;
    for(auto d:disp) {
        int r_new = r+d.first;
        int c_new = c+d.second;
        if(r_new>=0 && r_new<matrix.size() && c_new>=0 && c_new<matrix[0].size()) {
            if(matrix[r_new][c_new]>matrix[r][c] ) {
                int dist = 1+dfs(r_new,c_new,dp,matrix);
                ans = max(ans,dist);
            }
        }
    }
    if (ans==-1) {
        return dp[r][c]=1;
    }
    return dp[r][c]=ans;
    }
    int longestIncreasingPath(vector<vector<int>>& matrix) {
     int m = matrix.size();
     int n = matrix[0].size();
     vector<vector<int>> dp(m+1,vector<int>(n+1,-1));
     int ans=-1;
     for(int i=0;i<m;i++) {
        for(int j=0;j<n;j++) {
            if(dp[i][j]==-1) dfs(i,j,dp,matrix);
            ans = max(ans,dp[i][j]);
        }
     }   
     return ans;
    }
};

File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.py
class Solution:
    
        


    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        m = len(matrix[0])
        dp = [[-1 for _ in range(m+10)] for _ in range(n+10)]
        ans=0


        def dfs(i,j):
            if dp[i][j]!=-1: return dp[i][j]
            dirs = [(-1,0),(1,0),(0,1),(0,-1)]
            ans=-1
            for (di,dj) in dirs:
                iNew = i+di
                jNew = j+dj
                if(iNew>=0 and jNew>=0 and iNew<len(matrix) and jNew<len(matrix[0])):
                    if matrix[iNew][jNew]>matrix[i][j]:
                        ans = max(ans,1+dfs(iNew,jNew))
            if ans==-1:
                dp[i][j]=1
            else:
                dp[i][j]=ans
            return dp[i][j]


        for i in range(n):
            for j in range(m):
                if dp[i][j]==-1:
                    ans = max(ans,dfs(i,j))
        return ans
                    
        